### PR TITLE
Limit peer bandwidth

### PR DESCRIPTION
**Pull request by JakubFornadel**
## Purpose

!!! Postponed - need to get long term statistics from our and community nodes. Can be done by checking the logs or rpc call, see https://github.com/Taraxa-project/taraxa-node/issues/1505 

Implemented generic protection against ddos by limiting allowed bandwidth per node. We should be able to protect against:

- sending too many packets of the same type -> more than 10k per 10s time period
- sending packets with too big overall size of the same type -> more than 10MB per 10s time period
- sending too many packets (all types) -> more than 25k per 10s time period
- sending packets with too big overall size of packets(all types) -> more than 25MB per 10s time period

## For reviewers

- please comment out this PR with the bandwidths limits that you think make sense
- until we implement weighted votes, we will have to set these limits according to votes (could be few hundred k every 10 sec) so not ideal....

## Measurements
I did measure periodically different time periods and create statistics based on that:

#### 5 sec time period
```
[Abs max all count -> total_count: 18851, total_size: 5081421 [B], avg_size: 269 [B], processing_duration: 7144732 [us], avg processing_duration: 379 [us], tp_wait_duration: 32923525 [us], avg tp_wait_duration: 1746 [us]]
[Abs max all size -> total_count: 90, total_size: 13829584 [B], avg_size: 153662 [B], processing_duration: 1276634 [us], avg processing_duration: 14184 [us], tp_wait_duration: 1900286 [us], avg tp_wait_duration: 21114 [us]]
[Abs max single count: VotePacket -> total_count: 13666, total_size: 3156846 [B], avg_size: 231 [B], processing_duration: 3833783 [us], avg processing_duration: 280 [us], tp_wait_duration: 207563669 [us], avg tp_wait_duration: 15188 [us]]
[Abs max single size: PbftSyncPacket -> total_count: 84, total_size: 13829482 [B], avg_size: 164636 [B], processing_duration: 1276417 [us], avg processing_duration: 15195 [us], tp_wait_duration: 1899680 [us], avg tp_wait_duration: 22615 [us]]
```

#### 10 sec time period
```
[Abs max all count -> total_count: 29921, total_size: 7961053 [B], avg_size: 266 [B], processing_duration: 14056448 [us], avg processing_duration: 469 [us], tp_wait_duration: 166729701 [us], avg tp_wait_duration: 5572 [us]]
[Abs max all size -> total_count: 147, total_size: 20781498 [B], avg_size: 141370 [B], processing_duration: 2718515 [us], avg processing_duration: 18493 [us], tp_wait_duration: 8043970 [us], avg tp_wait_duration: 54720 [us]]
[Abs max single count: VotePacket -> total_count: 22344, total_size: 5161464 [B], avg_size: 231 [B], processing_duration: 9363095 [us], avg processing_duration: 419 [us], tp_wait_duration: 292517630 [us], avg tp_wait_duration: 13091 [us]]
[Abs max single size: PbftSyncPacket -> total_count: 136, total_size: 20781311 [B], avg_size: 152803 [B], processing_duration: 2718056 [us], avg processing_duration: 19985 [us], tp_wait_duration: 8042559 [us], avg tp_wait_duration: 59136 [us]]

```
#### 30 sec time period
```
[Abs max all count -> total_count: 80974, total_size: 21501725 [B], avg_size: 265 [B], processing_duration: 32576973 [us], avg processing_duration: 402 [us], tp_wait_duration: 213472696 [us], avg tp_wait_duration: 2636 [us]]
[Abs max all size -> total_count: 365, total_size: 46063263 [B], avg_size: 126200 [B], processing_duration: 6330972 [us], avg processing_duration: 17345 [us], tp_wait_duration: 9325279 [us], avg tp_wait_duration: 25548 [us]]
[Abs max single count: VotePacket -> total_count: 57574, total_size: 13299594 [B], avg_size: 231 [B], processing_duration: 19360155 [us], avg processing_duration: 336 [us], tp_wait_duration: 176083979 [us], avg tp_wait_duration: 3058 [us]]
[Abs max single size: PbftSyncPacket -> total_count: 328, total_size: 46062634 [B], avg_size: 140434 [B], processing_duration: 6329073 [us], avg processing_duration: 19295 [us], tp_wait_duration: 9320363 [us], avg tp_wait_duration: 28415 [us]]

```
#### 60 sec time period
```
[Abs max all count -> total_count: 155734, total_size: 41653804 [B], avg_size: 267 [B], processing_duration: 69653165 [us], avg processing_duration: 447 [us], tp_wait_duration: 690837873 [us], avg tp_wait_duration: 4436 [us]]
[Abs max all size -> total_count: 702, total_size: 86727110 [B], avg_size: 123542 [B], processing_duration: 9846762 [us], avg processing_duration: 14026 [us], tp_wait_duration: 15793166 [us], avg tp_wait_duration: 22497 [us]]
[Abs max single count: VotePacket -> total_count: 109520, total_size: 25299120 [B], avg_size: 231 [B], processing_duration: 41415338 [us], avg processing_duration: 378 [us], tp_wait_duration: 608612007 [us], avg tp_wait_duration: 5557 [us]]
[Abs max single size: PbftSyncPacket -> total_count: 630, total_size: 86725886 [B], avg_size: 137660 [B], processing_duration: 9843540 [us], avg processing_duration: 15624 [us], tp_wait_duration: 15784464 [us], avg tp_wait_duration: 25054 [us]]
```

### CODE REVIEW

File: libraries/cli/include/cli/default_config.hpp [hunk 1]
```diff
@@ -8,14 +8,17 @@ const char *default_json = R"foo({
...
   "network_ideal_peer_count": 10,
   "network_max_peer_count": 50,
   "network_sync_level_size": 10,
   "network_packets_processing_threads": 10,
   "deep_syncing_threshold" : 10,
+  "bandwidth_throttle_period_seconds" : 10,
+  "max_allowed_total_packets_size" : 52428800,
+  "max_allowed_total_packets_count" : 500,
+  "max_allowed_same_type_packets_size" : 10485760,
+  "max_allowed_same_type_packets_count" : 100,
```
Comment (MatusKysel) [inline libraries/cli/include/cli/default_config.hpp:21]: I find it too low, we have now millions of unprocessed vote packets

Comment (JakubFornadel) [reply libraries/cli/include/cli/default_config.hpp]: yeah I dont really know what limits should I set there, currently it is (per 10sec):

- 10k packets count of the same type
- 10MB packets size of the same type
- 25k packets count of the all type
- 15MB packets size of the all type

File: libraries/cli/include/cli/testnet_config.hpp [hunk 2]
```diff
@@ -8,14 +8,17 @@ const char *testnet_json = R"foo({
   "network_address": "0.0.0.0",
   "network_tcp_port": 10002,
   "network_udp_port": 10002,
-  "network_simulated_delay": 0,
-  "network_transaction_interval": 100,
-  "network_bandwidth": 40,
+  "network_transaction_interval": 100
```
Comment (gaoqi8649) [inline libraries/cli/include/cli/testnet_config.hpp:11]: missing "," in end

**Review by gaoqi8649 — COMMENTED** [review]
I think first should discuss what's the TPS will target to support?

**Comment by JakubFornadel** [comment]
> I think first should discuss what's the TPS will target to support?

File: libraries/core_libs/network/src/tarcap/stats/bandwidth_stats.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        std::chrono::steady_clock::now() - node_stats.bandwidth_throttle_period_start_);
+
+    // period duration passed, reset stats to zero
+    if (current_period_duration > k_bandwidth_throttle_period_duration_) {
+      node_stats = NodeStats();
+    }
+
+    // Adjust node bandwidth stats
+    total_packets_count = ++node_stats.total_packets_count_;
+    total_packets_size = node_stats.total_packets_size_ += packet_size;
```
Comment (JakubFornadel) [inline libraries/core_libs/network/src/tarcap/stats/bandwidth_stats.cpp:33]: bug +=

File: libraries/core_libs/network/src/tarcap/stats/bandwidth_stats.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    auto& packet_type_stats = node_stats.packets_types_stats_[packet_type];
+    type_packets_count = ++packet_type_stats.packets_count_;
+    type_packets_size = packet_type_stats.packets_size_ += packet_size;
```
Comment (JakubFornadel) [inline libraries/core_libs/network/src/tarcap/stats/bandwidth_stats.cpp:37]: bug +=

**Pull request closed without merging**